### PR TITLE
metro scheduler implementation

### PR DESCRIPTION
**Pull request by sunnyagg**
- Adds nodebinding model
- watch handlers for node, subscription, binding changes
- Scheduler implementation (Load Based, Random)
- Supporting Registry, Repo, Core APIs

### CODE REVIEW

File: service/worker/service.go [hunk 1]
```diff
@@ -245,14 +266,14 @@ func (svc *Service) lead(ctx context.Context) error {
...
 		return err
 	}
 
 	// create subscription cache
 	logger.Ctx(ctx).Infof("getting active subscriptions for cache")
-	svc.subCache, err = svc.subscriptionCore.ListKeys(svc.ctx)
+	svc.subCache, err = svc.subscriptionCore.List(svc.ctx, "")
 	if err != nil {
 		return err
 	}
```
Comment (Abhishekvrshny) [inline service/worker/service.go:279]: I think there can be a race condition here, i.e a node or subs can be added or deleted between this step and when the actual watch is set. Can we update these caches as part of setting the watch itself?

File: service/worker/service.go [hunk 2]
```diff
@@ -327,68 +346,169 @@ func (svc *Service) stepDown() {
...
+			if old.Key() == new.Key() {
 				found = true
 				break
 			}
 		}
 
 		if !found {
-			logger.Ctx(svc.ctx).Infow("node added", "node_key", new)
+			logger.Ctx(svc.ctx).Infow("node added", "key", new.Key())
+			// Do nothing for now
```
Comment (Abhishekvrshny) [inline service/worker/service.go:511]: This should also trigger scheduler right?

File: service/worker/service.go [hunk 3]
```diff
@@ -280,8 +300,7 @@ func (svc *Service) lead(ctx context.Context) error {
 		WatchPath: "metro/subscriptions",
 		Handler: func(ctx context.Context, pairs []registry.Pair) {
 			logger.Ctx(ctx).Infow("nodes watch handler data", "pairs", pairs)
-			newSubs := registry.GetKeys(pairs)
-			svc.handleSubUpdates(newSubs)
+			svc.handleSubUpdates(pairs)
```
Comment (Abhishekvrshny) [inline service/worker/service.go:303]: `handleNodeUpdates ` and `handleSubUpdates ` happens concurrently. Do you see any side-effects? Does this need to be synchronized?

File: service/worker/service.go (hunk 1 above)
Comment (sunnyagg) [reply service/worker/service.go:284]: when we create watch, that gives a mandatory first callback with all data present under that key prefix, i guess we can then just remove the manual fetch and depend only on the watch, as first watch will update the empty cache

File: service/worker/service.go (hunk 2 above)
Comment (sunnyagg) [reply service/worker/service.go:535]: for now, rebalance is not implemented, we dont need to change the current assignment for node additions

File: service/worker/service.go (hunk 3 above)
Comment (sunnyagg) [reply service/worker/service.go]: concurrently is fine, issue is with the failures in both cases, as watch is triggered only once, we need some intermediate storage if processing fails

i.e. if a subscription scheduling fails, put that in some channel and process it in separate go routine

**Review by Abhishekvrshny — APPROVED** [review]

**Pull request merged**